src/google/protobuf/compiler/objectivec/objectivec_file.cc

```diff
@@ -414,9 +414,28 @@ void FileGenerator::GenerateSource(io::Printer *printer) {
     printer->Print(
         "#pragma clang diagnostic ignored \"-Wdirect-ivar-access\"\n");
   }
-
   printer->Print(
-      "\n"
+		"\n");
+  std::set<string> fwd_decls;
+  for (const auto& generator : message_generators_) {
+    generator->DetermineStaticForwardDeclarations(&fwd_decls);
+  }
+  if (fwd_decls.size() > 0) {
+		printer->Print(
+        "#pragma mark - Objective C Class references\n"
+        "// This somewhat arcane code forces linkage of classes from static archives by\n"
+        "// adding a concrete reference to the classes.\n"
+        "// We don't use `[Foo class]` because we need a static value for our initializer.\n"
+        "// This also has the added benefit of reducing size in that we don't have to\n"
+        "// encode the class names and look them up at runtime.\n");
+  }
+  for (const auto& i : fwd_decls) {
+    printer->Print("$value$;\n", "value", i);
```

Comment:
This flow seems a little odd/cryptic.

You call something to collect a set of things, that the name implies are some sorta forward decl, but when you print them, you know to add the `;` on the end.  It almost seems like they should be complete, i.e. - maybe just add the newline, but the ending semi should be within them, or you should only be collect the names of the classes, and the full formatting of what to do with the name should exist here.

also, by just collecting the names, each leaf doing an override wouldn't have to know the asm magic, they would just have to list the class, and the asm magic would all be in one place.

Reply:
Fixed somewhat :)

I will centralize the asm magic in a later change that includes the extensions. I fixed it so that the semi is included in them.